File: atcoder/codefestival/tyama_atcodercodefestival2014finalJ_TLE.cpp
#include <unordered_set>
#include <map>
#include <deque>
#include <vector>
#include <algorithm>
#include <cstdio>
using namespace std;
typedef pair<long long,long long> pll;
int main(){
	long long first,second,k;
	scanf("%lld%lld%lld",&first,&second,&k);
	unordered_set<long long>lst={0};
	map<pll,long long>depth={{{0,0},0}};
	deque<pll>q={{0,0}};
	for(;!q.empty();){
		pll x=q.front();
		q.pop_front();
		vector<pll>a={
			{first,x.second},
			{x.first,second},
			{0,x.second},
			{x.first,0},
			{x.first-min(x.first,second-x.second),x.second+min(x.first,second-x.second)},
			{x.first+min(first-x.first,x.second),x.second-min(first-x.first,x.second)},
		};
		for(auto &e:a){
			if(depth.find(e)==depth.end() && depth[x]<k){
				lst.insert(e.first);
				lst.insert(e.second);
				depth[e]=depth[x]+1;
				q.push_back(e);
			}
		}
	}
	printf("%d\n",lst.size());
	return 0;
}

File: atcoder/codefestival/tyama_atcodercodefestival2014finalJ_TLE.py
#!/usr/bin/python
def checkio(first, second, k):
	lst={0}
	depth={(0,0):0}
	q=[(0,0)]
	while len(q)>0:
		x=q[0]
		q.pop(0)
		a=[
			('01',(first,x[1])),
			('02',(x[0],second)),
			('10',(0,x[1])),
			('20',(x[0],0)),
			('12',(x[0]-min(x[0],second-x[1]),x[1]+min(x[0],second-x[1]))),
			('21',(x[0]+min(first-x[0],x[1]),x[1]-min(first-x[0],x[1]))),
		]
		for e in a:
			if e[1] not in depth and depth[x]<k:
				lst.add(e[1][0])
				lst.add(e[1][1])
				depth[e[1]]=depth[x]+1
				q.append(e[1])
	return len(lst)

a=[int(e) for e in raw_input().split()]
print(checkio(*a))